replace our usage of pexpect in IsolatedManager with ansible-runner

ryanpetrello opened the pull request:
This is the _last_ major bit of runner-related cleanup that I know of - I can't think of any other place in the codebase where we're doing artisanal `pexpect` calls 🎉

ryanpetrello committed:
replace our usage of pexpect in IsolatedManager with ansible-runner

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -20,143 +19,83 @@
         return os.path.join(self.private_data_dir, *args)
 
+    def run_management_playbook(self, playbook, private_data_dir, **kw):
+        iso_dir = tempfile.mkdtemp(
```
Run each management playbook in a distinct tmpdir _within_ `private_data_dir`.

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -319,58 +248,51 @@ def health_check(cls, instance_qs, awx_application_version):
-        args.extend(['--forks', str(len(instance_qs))])
-        env = cls._base_management_env()
+        # TODO: runner doesn't have a --forks arg
+        #args.extend(['--forks', str(len(instance_qs))])
```
ansible-runner doesn't support this in its programmatic interface, so I've got to figure out a way to pass it in.

ryanpetrello replied:
Talked to @chrismeyersfsu; I'll do this later when it's supported in the next ansible-runner version:

https://github.com/ansible/ansible-runner/pull/241

The default `--forks` is probably fine for now; in practice, people don't have isolated clusters with more than maybe 10-20 nodes per controller.

matburt replied:
This is now merged

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -319,58 +248,51 @@ def health_check(cls, instance_qs, awx_application_version):
+                for instance in instance_qs:
+                    task_result = {}
+                    try:
+                        task_result = runner_obj.get_fact_cache(instance.hostname)
```
@matburt wow this interface is so much nicer.

ryanpetrello commented on awx/main/expect/run.py:
```diff
@@ -1,316 +0,0 @@
-#! /usr/bin/env python
```
🔪🔪🔪🔪🔪

ryanpetrello commented on awx/main/management/commands/test_isolated_connection.py:
```diff
@@ -25,23 +23,21 @@ def handle(self, *args, **options):
-                args = run.wrap_args_with_ssh_agent(args, ssh_key_path, ssh_auth_sock)
-            try:
-                print(' '.join(args))
-                subprocess.check_call(args)
```
🔪🔪🔪🔪🔪

ryanpetrello commented on awx/main/tests/functional/test_tasks.py:
```diff
@@ -126,23 +126,6 @@ def old_version(self, control_group):
         return inst
 
-    @pytest.mark.skip(reason='fix after runner merge')
-    def test_old_version(self, control_instance, old_version):
```
We don't do version comparisons between normal AWX nodes and isolated ones because the isolated ones are just *runner* now, not some special awx version.

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -20,143 +19,83 @@
-        ))
+        def finished_callback(runner_obj):
+            if runner_obj.status == 'failed':
+                stdout = runner_obj.stdout.read()
```
@AlanCoding if a management playbook (any of them) fails, do `playbook_logger.error(stdout)` and also send a verbose event so we can see the failure at the `stdout` endpoint.

AlanCoding replied:
We did this same thing before moving to runner (removed line 128). It's somewhat gory to get management playbook details in the job stdout, but when it can't be started, that instant feedback is really important.

The only difference I see is that the EOF event isn't sent, which I assume runner does. So this looks pretty good for just keeping on doing what it used to do.

ryanpetrello replied:
@AlanCoding I talked to @chrismeyersfsu and decided to remove the verbose event because we're worried it's going to mess w/ the event counting `event_processing_finished` mechanism.  If these playbooks fail, check the management playbook logs for error info.

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -20,143 +19,83 @@
+                event_data = {'event': 'verbose', 'stdout': stdout, self.event_data_key: self.instance.id}
+                CallbackQueueDispatcher().dispatch(event_data)
+            else:
+                playbook_logger.info(runner_obj.stdout.read())
```
This is also kind of nice for development purposes to see the playbook output go by.

chrismeyersfsu replied:
HOT

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -20,143 +19,83 @@
+            getattr(settings, 'AWX_ISOLATED_KEY_GENERATION', False) is True,
+            getattr(settings, 'AWX_ISOLATED_PRIVATE_KEY', None)
+        ]):
+            params['ssh_key'] = settings.AWX_ISOLATED_PRIVATE_KEY
```
Pass in the controller -> iso RSA key (if any)

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -191,22 +147,12 @@ def check(self, interval=None):
-        job_timeout = remaining = self.job_timeout
         dispatcher = CallbackQueueDispatcher()
         while status == 'failed':
-            if job_timeout != 0:
```
We don't have to track timeouts at all anymore - runner does this for us.

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -273,30 +212,21 @@ def check(self, interval=None):
-            logger.warning('Isolated job {} cleanup error, output:\n{}'.format(self.instance.id, output))
 
     @classmethod
-    def update_capacity(cls, instance, task_result, awx_application_version):
```
unused var

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -308,8 +238,7 @@ def update_capacity(cls, instance, task_result, awx_application_version):
         instance.save(update_fields=['cpu_capacity', 'mem_capacity', 'capacity', 'version', 'modified'])
 
-    @classmethod
-    def health_check(cls, instance_qs, awx_application_version):
```
unused var

ryanpetrello committed:
set PYTHONPATH properly for isolated usage of runner

ryanpetrello committed:
remove main.expect tests (this functionality exists in runner now)

ryanpetrello commented:
@matburt @AlanCoding @chrismeyersfsu @jbradberry after this passes tests and goes in, I want to write more unit tests for `awx.main.expect.isolated_manager`.

Also, because it's the only thing in `awx.main.expect` (there's no more `pexpect` usage), how would you feel about (in another PR):

`git mv awx/main/expect awx/main/isolated`
`git mv awx/main/isolated/isolated_manager.py awx/main/isolated/manager.py`

chrismeyersfsu commented on awx/main/expect/isolated_manager.py:
```diff
@@ -1,162 +1,112 @@
     def path_to(self, *args):
         return os.path.join(self.private_data_dir, *args)
 
+    def run_management_playbook(self, playbook, private_data_dir, **kw):
```
Can we list all the explicitly list all the kwargs that are expected?

ryanpetrello replied:
eeeeeeeh?  This could be a _really_ long list, and it's subject to change/grow with each version of runner:

https://github.com/ansible/ansible-runner/blob/master/ansible_runner/interface.py#L74

chrismeyersfsu commented on awx/main/expect/isolated_manager.py:
```diff
@@ -319,58 +257,51 @@ def health_check(cls, instance_qs, awx_application_version):
-                idle_timeout=timeout, job_timeout=timeout,
-                pexpect_timeout=5
+            private_data_dir = tempfile.mkdtemp(
+                prefix='awx_iso_heartbeat_',
```
good, this makes it easier to find. Any other contextual data we might want to include would be helpful.

ryanpetrello committed:
send an EOF event if isolated dispatch fails

ryanpetrello committed:
clean up old usage of `idle_timeout`

cmeyers and I looked at this and can't tell where/why/how you'd actually
set this setting - it looks like really old ~2014-2015 Tower history
that probably isn't actually in use

ryanpetrello commented on awx/main/expect/isolated_manager.py:
```diff
@@ -393,10 +335,18 @@ def run(self, instance, private_data_dir, playbook, module, module_args,
         status, rc = self.dispatch(playbook, module, module_args)
         if status == 'successful':
             status, rc = self.check()
+        else:
```
If the initial controller -> isolated dispatch fails, send an EOF event so the UI is happy.
